src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs

```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
 
-            BoundExpression result = (temps.Count == 0 && stores.Count == 0) ?
+                return (temps.Count == 0 && stores.Count == 0) ?
                     conditionalExpression :
                     new BoundSequence(
                         syntax,
-                        temps.ToImmutable(),
-                        stores.ToImmutable(),
+                        temps.ToImmutableAndFree(),
+                        stores.ToImmutableAndFree(),
                         conditionalExpression,
                         conditionalExpression.Type);
+            }
 
-            temps.Free();
-            stores.Free();
+            // Rewrites the null coalescing operator in the case where the result type is the underlying
+            // non-nullable value type of the left side
+            BoundExpression rewriteNullCoalescingAssignmentForValueType()
+            {
+                Debug.Assert(node.LeftOperand.Type.IsNullableType());
+                Debug.Assert(node.Type.Equals(node.RightOperand.Type));
 
-            return result;
+                // We lower the expression to this form:
+                //
+                // var tmp = lhsRead.GetValueOrDefault();
+                // if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
+                // tmp
+                //
+                // Except that a is only evaluated once.
+
+                var leftOperand = node.LeftOperand;
+                if (!TryGetNullableMethod(leftOperand.Syntax,
+                                          leftOperand.Type,
+                                          SpecialMember.System_Nullable_T_GetValueOrDefault,
+                                          out var getValueOrDefault))
+                {
+                    return BadExpression(node);
+                }
+
+                if (!TryGetNullableMethod(leftOperand.Syntax,
+                                          leftOperand.Type,
+                                          SpecialMember.System_Nullable_T_get_HasValue,
+                                          out var hasValue))
+                {
+                    return BadExpression(node);
+                }
+
+                // If null coalescing assignment is supported in expression trees, the below code
+                // will need to be updated to support property accesses as well as calls. Currently,
+                // MakeRValue will never return a BoundPropertyAccess except in expression trees.
+                Debug.Assert(!_inExpressionLambda && lhsRead.Kind != BoundKind.PropertyAccess);
+
+                // If lhsRead is a call, such as to a property accessor, save the result off to a temp. This doesn't affect
+                // the standard ??= case because it only uses lhsRead once.
+                if (lhsRead.Kind == BoundKind.Call)
+                {
+                    var lhsTemp = _factory.StoreToTemp(lhsRead, out var store, kind: SynthesizedLocalKind.Spill);
```

Comment:
Why do we need `SynthesizedLocalKind.Spill`?

Reply:
When we get to the SpillSequenceSpiller, it asserts that all synthesized locals are long lived. The default (`LoweringTemp`) is not.

---
In reply to: [293111279](https://github.com/dotnet/roslyn/pull/36329#discussion_r293111279) [](ancestors = 293111279)

Reply:
I see no good reason why we need this temporary to be long lived, since the other ones used in compound assignments are not.

Reply:
The other ones are not used in a spill sequence, this is.

---
In reply to: [293135135](https://github.com/dotnet/roslyn/pull/36329#discussion_r293135135) [](ancestors = 293135135)